Handling long list of arguments for extractLocStrings

csknns opened the pull request:
Handling long list of arguments for extractLocStrings by writing them in a arguments plist file. 

This solves errors that causes crash with Exception: 'Couldn't posix_spawn: error 7'.

Fixes Flinesoft/BartyCrouch#92
This PR with minor modifications can also fix Flinesoft/BartyCrouch#99

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -28,7 +28,9 @@ public class CodeCommander {
         let customFunctionArgs = customFunction != nil ? ["-s", "\(customFunction!)"] : []
 
-        let arguments = ["extractLocStrings"] + files + ["-o", stringsFilePath] + customFunctionArgs + ["-q"]
+        let argumentsWithOutTheFiles = ["extractLocStrings"] + ["-o", stringsFilePath] + customFunctionArgs + ["-q"]
```
Please fix the typo here ate `withOut` making the `O` lowercase.

Dschee replied:
Also, please put the `["-o", ...]` array items into the first array since there's no `files` anymore.

csknns replied:
@Dschee I am not sure where to place the the `["-o", ...]` array items, in the `files` array or the `customFunctionArgs`? The files array I am handling it as a list of files and not necessary has a list of arguments. If the files list is small the files are used as arguments, else the argument is the argument plist file.
I think the -o will always be in the argument list, regardless how we pass the list files, so it seems to be rightly placed.

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+/// Code to handle argument list extractLocStrings exceeding maximum supported character length of
+/// arguments. See https://github.com/Flinesoft/BartyCrouch/issues/92
+extension CodeCommander {
+
```
Please remove this empty line.

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+
+    /// extractLocStrings tools supports instead of passing files as arguments in the command line
+    /// to pass a plist with the files. This is the format of that plist.
+    struct ArgumentsPlist: Encodable {
```
Could you please extract the Plist, handling and conversions into separate file(s)? Thank you!

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+        let completeArgumentList = existingArguments + files
+
+        if isArgumentListTooLong(completeArgumentList) {
+            //If the argument list gets to long we write the files in a plist and pass that as
```
Please always add a whitespace after any `//` for comment lines.

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+        }
+    }
+
+    //Method to determent if the argument list is to long to pass to new process
```
See above.

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+        // Get maximum length of arguments for a new process (in Bytes)
+        // See https://www.in-ulm.de/~mascheck/various/argmax/
+        let getconfCommand = "expr `getconf ARG_MAX` - `env|wc -c` - `env|wc -l` \\* 4 - 2048"
+        guard let commandResult = try? Task.capture(bash: getconfCommand),
```
Please use this format instead:

```swift
guard
    let commandResult = ...,
    let maxArgumentLength = ...
else {
    return false
}
```

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+    private func createTemporaryArgumentsPlistFile() -> URL {
+        let temporaryFilename = ProcessInfo().globallyUniqueString
+        var temporaryPath = URL(fileURLWithPath: NSTemporaryDirectory(),
+                                isDirectory: true)
```
Please keep this in the line above. Up to 160 characters all is fine.

Dschee commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -59,3 +61,73 @@ public class CodeCommander {
+        temporaryPath.appendPathExtension("plist")
+
+        return temporaryPath
+
```
Please remove this empty line.

csknns commented:
I will apply your suggestion and update this PR. Since most of the commends are regarding source code style, you consider running swiftlint as part of the CI? It would reduce time you take to review the PR.

Regarding #99, the fix is small. I just need to add one more condition(if total arguments > _POSIX_ARG_MAX i.e. 4096) to determent if the arguments plist file should be used instead of passing the file list in the command line.

csknns commented:
I can also add some unit tests regarding the argument plist file generation. But keep in mind that there is not documentation that I could find regarding the format of the file, so I pretty much reserve engineered it's structure.

Dschee commented:
Sure, that's fine with me. Thank you very much!

csknns commented:
@Dschee I have update the PR, take a look when you can :)

aadjemonkeyrock commented:
Any reasons why this hasn't been processed yet? I'm running into the same issue and would like to see this completed.

Jeehut commented:
@aadjemonkeyrock I didn't have time to code review this yet. I would like to see this completed as well, but I need to test it to make sure no one gets affected by any unexpected behavior as many are relying on BartyCrouch to just work.

If this feature was added as an option which wouldn't affect any existing users, I could merge without much testing, but it currently is changing even the behavior with default settings. (In that case the README would need and update as well, of course.

@csknns I just had a look again and found multiple things that I would like to have changed:

1. Please add a CHANGELOG entry that describes this change.
2. Please check if you find any warnings when running `swiftlint` (with the latest version)
3. Please use a line length of 160 and prevent any unnecessary breaks, e.g. in comments

aadjemonkeyrock commented:
@Jeehut I would be more than willing to do some tests, but I've no idea how to install the forked code. 

Anyway, I appreciate your feedback and good work. Don't rush on my part.

Jeehut commented:
@aadjemonkeyrock BartyCrouch is just a Swift package, so you can build it the way you can build any SwiftPM project, see the documentation here: https://github.com/apple/swift-package-manager/blob/master/Documentation/Usage.md#creating-an-executable-package

In general, if you want to know how to build a project that is distributed via Homebrew, you can always look up how it's built by Homebrew by opening it's [formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/bartycrouch.rb).

For BartyCrouch specifically, we have a `Makefile` which simplifies installing it locally. You just need to run `make install`. :)

I hope that helps.

csknns commented:
@Jeehut I have added an entry in the Changelog, the swiftlint only reported a false positive in the tests. I have also updated the comments to use the 160 limit(did not find and place other than comments)

@aadjemonkeyrock It would be very useful to have more people test this! You can checkout the code with the fix, build it, and run the bartycrouch binary from the that build instead of the version installed in your system.
1. Checkout the code from my fork https://github.com/csknns/BartyCrouch, when switch to the branch with the fix csknns:cknns-stable
2. To build it as @Jeehut said you can simple open a terminal and run `swift build` in the repo top folder
3. The above command will produce an binary in the folder `.build` (this is by default hidden), the path of the binary is `.build/debug/bartycrouch`, so you can use the full path of the binary to run it in your project and check if it works. For example `/Users/auser/BartyCrouch/.build/debug/bartycrouch update`

aadjemonkeyrock commented:
Sorry for the late reply, but I've been a bit too busy. 

I followed the instructions but I do have some problems. Below I paste the complete output of bartycrouch update

Aads-MacBook-Pro-2:gw.ios aadje$ /Users/aadje/Documents/workspace/BartyCrouch/.build/debug/bartycrouch update
Starting Task 'Update Interfaces' ...
2019-11-14 10:24:45.576: ✅  ./gw.ios/Resources/Base.lproj/LaunchScreen.storyboard:  Successfully updated strings file(s) of Storyboard or XIB file.
2019-11-14 10:24:49.047: ✅  ./gw.ios/Resources/Base.lproj/Startup.storyboard:  Successfully updated strings file(s) of Storyboard or XIB file.
2019-11-14 10:24:49.516: ✅  ./gw.ios/Resources/Base.lproj/Main.storyboard:  Successfully updated strings file(s) of Storyboard or XIB file.
2019-11-14 10:24:49.587: ✅  ./gw.ios/Resources/Base.lproj/Settings.storyboard:  Successfully updated strings file(s) of Storyboard or XIB file.
Task 'Update Interfaces' took 7.716 seconds.
Starting Task 'Update Code' ...
2019-11-14 10:24:49.916 bartycrouch[67604:5628633] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Couldn't posix_spawn: error 7'
*** First throw call stack:
(
        0   CoreFoundation                      0x00007fff392e2f53 __exceptionPreprocess + 250
        1   libobjc.A.dylib                     0x00007fff6f3a8835 objc_exception_throw + 48
        2   Foundation                          0x00007fff3b91bfd5 -[NSConcreteTask launchWithDictionary:error:] + 5014
        3   bartycrouch                         0x0000000104e55179 $s8SwiftCLI4TaskC6launch030_19E0ACB72F1C972020BFBD69850F9J1FLLyyF + 569
        4   bartycrouch                         0x0000000104e54a8f $s8SwiftCLI4TaskC7runSyncs5Int32VyF + 31
        5   bartycrouch                         0x0000000104e55b99 $s8SwiftCLI4TaskC3run_9arguments9directoryySS_SaySSGSSSgtKFZ + 441
        6   bartycrouch                         0x0000000104dd0547 $s14BartyCrouchKit13CodeCommanderC6export18stringsFilesToPath04fromd11InDirectoryJ014customFunctionySS_S2Abort trap: 6
Aads-MacBook-Pro-2:gw.ios aadje$

csknns commented:
@aadjemonkeyrock Thank you for taking the time to test the issue! Can you double check that you build the branch cknns-stable (git checkout cknns-stable before you build the project)? If yes can execute the following commands in the terminal and post here the results?
```expr `getconf ARG_MAX` - `env|wc -c` - `env|wc -l` \* 4 - 2048```
```expr `getconf _POSIX_ARG_MAX```

aadjemonkeyrock commented:
@csknns 

These are the results:

```
Aads-MacBook-Pro-2:BartyCrouch aadje$ git status
On branch cknns-stable
Your branch is up to date with 'origin/cknns-stable'.

nothing to commit, working tree clean
Aads-MacBook-Pro-2:BartyCrouch aadje$ expr `getconf ARG_MAX` - `env|wc -c` - `env|wc -l` \* 4 - 20
48
258895
Aads-MacBook-Pro-2:BartyCrouch aadje$ expr `getconf _POSIX_ARG_MAX`
4096
Aads-MacBook-Pro-2:BartyCrouch aadje$ 


```

csknns commented:
@aadjemonkeyrock I made a couple of changes, can you perform a git pull, rebuild and test again?
I have a project with many files that has the same error and my code change fixes the issue in my code base. If this is reproduced in other projects that I do not have access to debug, I will have to go with a interactive trial and error process to find the problem.

aadjemonkeyrock commented:
@csknns I've tested and for my project it now works as expected. 

However I do notice that also some sources in my /pods/ folder are parsed and end up in my localizedStrings. I'm not sure if this is supposed to happen

Jeehut commented on Sources/BartyCrouchKit/OldCommandLine/Extensions/TaskExtension.swift:
```diff
@@ -0,0 +1,31 @@
+    static func isArgumentListTooLong(_ argumentList: [String]) -> Bool {
+        // Get maximum length of arguments for a new process (in Bytes) and the maximum count for the list for arguments. See
+        // https://www.in-ulm.de/~mascheck/various/argmax/ for more info.
+//        guard
```
Why do we have code that is commented out here? Either remove or fix it, or at least add a comment why this should be kept that it's clear.

csknns replied:
I fixed the code, this detects if the argument list is to long and we need to use a plist argument file.

Jeehut commented on Sources/BartyCrouchKit/OldCommandLine/CodeCommander.swift:
```diff
@@ -28,7 +28,10 @@ public class CodeCommander {
+        let argumentsWithoutTheFiles = ["extractLocStrings"] + ["-o", stringsFilePath] + customFunctionArgs + ["-q"]
+
+        let arguments = try appendFiles(files, inListOfArguments: argumentsWithoutTheFiles)
+        print("command line arguments: " + arguments.joined(separator: " "))
```
What do you need this `print` for? If you want to output information for the user, please use the `print(_:level:)` method within BartyCrouch, not the plain `print` method. If this was for debugging purposes only, please remove it.

csknns replied:
I removed the print method

csknns commented:
@Jeehut 
1. I rebased to the latest main
2. Correct if the file argument list to the xcrun extractLocStrings is to big the command fails, so we write all the arguments (mostly the code files paths) to a plist file and we feed that instead as an argument.
It does not detect it automatically, I specify this with the -f flag. Check in file CodeCommander in line 78
3. I added the option that gives the ability to explicitly turn it on/off, for the user's convenience I leave the default mode as it is i.e. autodetect.

Jeehut commented on Tests/BartyCrouchKitTests/CommandLine/ExtractLocStringsTests.swift:
```diff
@@ -0,0 +1,29 @@
+
+// swiftlint:disable force_try
+
+class ExtractLocStringsTests: XCTestCase {
```
This entire file is not needed, in my opinion. See my general comment for rationale.

csknns commented:
@Jeehut Well my rational was to automatically detect if we need the new method (mostly for small performance gains) so in case my code has a bug the user will see it in case his project it too large, this is case BartyCrouch will fail anyway. Hence my justification to keep the auto detect on for the user's convenience. But I understand doing a gradual rollout starting with this feature turned of.

Also the reason I had the detection was that creating temp files and performing I/O, might introduce some small performance issues that I wanted to avoid If they are not needed.

Anyway, glad you find time to invest in BartyCrouch, keep up the good work!
